xla/service/collective_quantizer.cc

```diff
@@ -380,10 +386,20 @@ absl::StatusOr<bool> CollectiveQuantizer::Run(
 
   for (HloComputation* comp : module->MakeComputationPostOrder()) {
     for (HloInstruction* instr : comp->MakeInstructionPostOrder()) {
-      if (IsSupportedCollective(instr)) {
-        TF_RETURN_IF_ERROR(MatchDequantization(instr, &changed));
-        TF_RETURN_IF_ERROR(MatchQuantization(instr, &changed));
+      if (!IsSupportedCollective(instr)) {
+        continue;
+      }
+
+      // Either tries matching dequantization or quantization.
+      TF_ASSIGN_OR_RETURN(bool is_dequantized, MatchDequantization(instr));
+      changed |= is_dequantized;
+
+      if (is_dequantized) {
+        continue;
       }
+
+      TF_ASSIGN_OR_RETURN(bool is_quantized, MatchQuantization(instr));
+      changed |= is_quantized;
```

Comment:
So you mean that having slow/inefficient code is OK in XLA? 
The question here is that whether we can do both quantization and dequantization with the same collective instruction. If it can happen, it does seem quite interesting.

If it cannot happen, then you mean we just keep the slow code? The original code is not only slow but confusing because from a reader perspective, I don't really know whether you do see quantization and dequantization can happen at the same time. BTW, is there any test case for that behavior or have you observed anything like that in some case?

Reply:
Then things like this is also OK?
https://cs.opensource.google/tensorflow/tensorflow/+/master:third_party/xla/xla/service/collective_quantizer.cc;l=113?q=collective_quantizer

The replication analysis is expensive as it scans the entire HLO module. And the code here basically gives me one replication analysis per matched instructions so I end up with 20 in my code.

Reply:
I do not know of any realistic scenario where you would want to move both quantization and dequantization around the collective. But, it still doesn't hurt to check for both.

> So you mean that having slow/inefficient code is OK in XLA?

Slow/inefficient code is OK if the code is rarely run and doesn't make a significant impact on the overall performance. E.g. if a pass takes 3 milliseconds to run and could be optimized to take 1 millisecond at the cost of increased complexity, I don't think it's worth adding the complexity. In this case and in the case of running replication analysis, my hunch is that the performance gain is negligible, but performance numbers could persuade me otherwise.